packages/databyss-services/sources/actions.js

```diff
@@ -0,0 +1,87 @@
+import * as sources from './'
+import { SourceNotFoundError } from './SourceNotFoundError'
+
+import { GET_SOURCE, SET_SOURCE, CACHE_SOURCE } from './constants'
+
+export function fetchSource(id) {
+  return async dispatch => {
+    dispatch({
+      type: GET_SOURCE,
+      payload: {},
+    })
+
+    sources
+      .getSource(id)
+      .then(source => {
+        dispatch({
+          type: CACHE_SOURCE,
+          payload: { source, id },
+        })
+      })
+      .catch(() => {
+        dispatch({
+          type: CACHE_SOURCE,
+          payload: {
+            source: new SourceNotFoundError('Source not found', id),
+            id,
+          },
+        })
+      })
+  }
+}
+
+export function saveSource(sourceFields) {
+  return async dispatch => {
+    dispatch({
+      type: SET_SOURCE,
+      payload: {},
+    })
+
+    sources
+      .setSource(sourceFields)
+      .then(source => {
+        dispatch({
+          type: CACHE_SOURCE,
+          payload: { source, id: sourceFields._id },
+        })
+      })
+      .catch(() => {
+        dispatch({
+          type: CACHE_SOURCE,
+          payload: {
+            source: new SourceNotFoundError(
```

Comment:
At this point, we're not sure the error is a SourceNotFoundError.
In `request.js`, let's add an error class `ResourceNotFoundError` that we throw when the status is `404`. 
Then, we can just pass the error up:
```js
catch(err => {
   dispatch({
      payload: {
        source: err,
      }
   })
})
```